ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/ListSourceExecutor.java

```diff
@@ -208,13 +263,50 @@ private static SourceDescriptionWithWarnings describeSource(
         SourceDescriptionFactory.create(
             dataSource,
             extended,
-            getQueries(ksqlEngine, q -> q.getSourceNames().contains(dataSource.getName())),
-            getQueries(ksqlEngine, q -> q.getSinkName().equals(dataSource.getName())),
-            topicDescription
+            sourceQueries,
+            sinkQueries,
+            topicDescriptionOptional,
+            sourceConsumerOffsets
         )
     );
   }
 
+  private static List<SourceConsumerGroupOffset> consumerOffsets(
+      final TopicDescription topicDescription,
+      final Map<TopicPartition, ListOffsetsResultInfo> topicAndStartOffsets,
+      final Map<TopicPartition, ListOffsetsResultInfo> topicAndEndOffsets,
+      final Map<TopicPartition, OffsetAndMetadata> topicAndConsumerOffsets
+  ) {
+    final List<SourceConsumerGroupOffset> sourceConsumerGroupOffsets = new ArrayList<>();
+    for (TopicPartitionInfo topicPartitionInfo : topicDescription.partitions()) {
+      final TopicPartition tp = new TopicPartition(topicDescription.name(),
+          topicPartitionInfo.partition());
+      final ListOffsetsResultInfo startOffsetResultInfo = topicAndStartOffsets.get(tp);
+      final ListOffsetsResultInfo endOffsetResultInfo = topicAndEndOffsets.get(tp);
+      final OffsetAndMetadata offsetAndMetadata = topicAndConsumerOffsets.get(tp);
+      sourceConsumerGroupOffsets.add(
+          new SourceConsumerGroupOffset(
+              topicPartitionInfo.partition(),
+              startOffsetResultInfo != null ? startOffsetResultInfo.offset() : 0,
+              endOffsetResultInfo != null ? endOffsetResultInfo.offset() : 0,
+              offsetAndMetadata != null ? offsetAndMetadata.offset() : 0
+          ));
+    }
+    return sourceConsumerGroupOffsets;
+  }
+
+  private static String getServiceId(final KsqlConfig ksqlConfig) {
+    return ReservedInternalTopics.KSQL_INTERNAL_TOPIC_PREFIX
+        + ksqlConfig.getString(KsqlConfig.KSQL_SERVICE_ID_CONFIG);
+  }
```

Comment:
we should extract this to a utility class and reuse it in `QueryExecutor` (and ditto below)

Reply:
I've move it into `KsqlConfig` itself. Let me know if this is a better place.